funsor/joint.py

```diff
@@ -1,268 +1,24 @@
-        if gaussian_vars:
-            # Draw an expanded sample.
-            gaussian_sample_inputs = sample_inputs.copy()
-            for k, d in self.inputs.items():
-                if k not in result.gaussian.inputs and d.dtype != 'real':
-                    gaussian_sample_inputs[k] = d
-            gaussian = result.gaussian.unscaled_sample(gaussian_vars, gaussian_sample_inputs)
-            result = Joint(result.deltas, result.discrete) + gaussian
-        return result
+from funsor.cnf import Contraction, GaussianMixture, NullOp
+from funsor.delta import MultiDelta
+from funsor.domains import bint
+from funsor.gaussian import Gaussian, align_gaussian, cholesky_solve, cholesky_inverse
+from funsor.integrate import Integrate
+from funsor.ops import AssociativeOp
+from funsor.terms import Funsor, Number, Reduce, Unary, Variable, eager, moment_matching, normalize
+from funsor.torch import Tensor, align_tensor
 
 
-@eager.register(Joint, tuple, Funsor, Funsor)
-def eager_joint(deltas, discrete, gaussian):
-
-    if not isinstance(gaussian, (Number, Tensor, Gaussian)):
-        return Joint(deltas, discrete) + gaussian
-
-    if any(not isinstance(d, Delta) for d in deltas):
-        new_deltas = []
-        for d in deltas:
-            if isinstance(d, Delta):
-                new_deltas.append(d)
-            elif isinstance(d, (Number, Tensor)):
-                discrete += d
-            else:
-                raise ValueError("Invalid component for Joint: {}".format(d))
-        return Joint(tuple(new_deltas), discrete) + gaussian
-
-    if isinstance(gaussian, (Number, Tensor)) and gaussian is not Number(0):
-        discrete += gaussian
-        return Joint(deltas, discrete, Number(0))
-
-    # Demote a Joint to a simpler elementary funsor.
-    if not deltas:
-        if gaussian is Number(0):
-            return discrete
-        elif discrete is Number(0):
-            return gaussian
-    elif len(deltas) == 1:
-        if discrete is Number(0) and gaussian is Number(0):
-            return deltas[0]
-
-    return None  # defer to default implementation
-
-
-@eager.register(Independent, Joint, str, str, str)
-def eager_independent(joint, reals_var, bint_var, diag_var):
-    for i, delta in enumerate(joint.deltas):
-        if delta.name == diag_var:
-            delta = Independent(delta, reals_var, bint_var, diag_var)
-            deltas = joint.deltas[:i] + (delta,) + joint.deltas[1+i:]
-            discrete = joint.discrete
-            if bint_var in discrete.inputs:
-                discrete = discrete.reduce(ops.add, bint_var)
-            gaussian = joint.gaussian
-            if bint_var in gaussian.inputs:
-                gaussian = gaussian.reduce(ops.add, bint_var)
-            return Joint(deltas, discrete, gaussian)
-
-    return None  # defer to default implementation
-
-
-@dispatch(str, str, Variadic[Gaussian, Joint])
+@dispatch(str, str, Variadic[(Gaussian, GaussianMixture)])
```

Comment:
Is this pattern now too narrow?

Reply:
It shouldn't be, according to [this check in `master`](https://github.com/pyro-ppl/funsor/pull/157/files/8eb0930dcd1435f869cf51a66be50326974f0589#diff-48e50907b44e8bb24d6ee7485225e0ebL291)

Reply:
I think this might be caused by `normalize` being a bit too aggressive with pushing down `exp`, so that we get terms that look like `Contraction[..., Tuple[Unary[Exp, Gaussian], Tensor]]`

Reply:
Can we simply remove `Unary[Contraction]` simplification from `normalize`?

Reply:
Hmm, it might be easier to add `normalize` patterns that prevent oversimplification e.g.
```py
@normalize.register(Unary, ExpOp, GaussianMixture)
def normalize_unary_joint(op, arg):
    return None
```

Reply:
I guess I don't see why we'd want to move unary inside. Unary ops are pointwise and are generally cheaper to apply outside of reductions.

Reply:
> I guess I don't see why we'd want to move unary inside

The main use case here (as in `optimize` and `Contract`/`Integrate` already in `master`) is in rewriting integrals of large joint distributions.  Pushing the unary ops down to the leaves makes pattern matching easier, simplifies the normal form and lets us reuse the optimizer for integrals.  In these situations it's unlikely that unary ops would actually be evaluated, since they would eventually be rewritten back into `Integrate`s, KL divergences, etc.

I don't see a reason to change this unless we also want to completely decouple `Integrate` from `Contraction` (i.e. never rewrite `Integrate` to `Contraction` or vice versa), which I suppose we could discuss separately.